### PR TITLE
Shake&Tune v4.0.0

### PR DESCRIPTION
**Pull request by Frix-x**
# Shake&Tune v4.0.0

This new version is a complete rewrite of Shake&Tune to make it a real and proper Klipper plugin. While this doesn't change much in terms of usability, it's a lot for me, and it will help provide new features more integrated into Klipper in the future!

# Breaking changes

Since Shake&Tune now runs from within Klipper, **there are some changes that need to be made to the way it is installed on your machine**:
1. You will need to run the install script manually again using this command in SSH:
```bash
cd ~/klippain_shaketune && ./install.sh
```

2. Then you will need to update your moonraker.conf using this new template:
```
[update_manager Klippain-ShakeTune]
type: git_repo
origin: https://github.com/Frix-x/klippain-shaketune.git
path: ~/klippain_shaketune
virtualenv: ~/klippy-env
requirements: requirements.txt
system_dependencies: system-dependencies.json
primary_branch: main
managed_services: klipper
```

3. Then follow the new installation instructions to add the `[shaketune]` section to your printer.cfg. The old `[include ./K-ShakeTune/*.cfg]` can be safely removed!

# New Features

 * Shake&Tune runs inside Klipper and was fully rewritten in Python... No custom macros anymore!
 * The belt comparison macro has been updated to remove the spectrogram, which was not really useful. However, a new graph has been added to replace it, and this should give a good and easy overview of the belts' behavior and how much they match: if the plot line is in the green zone, everything is fine on your belt's side and you can continue to the next tool. Quick and easy!
 * Added the ability to test CoreXZ kinematics in the belt comparison tool to help you tension your belts on this type of machine. This is quite experimental and a new feature only possible because it's now a Klipper plugin. Note: The tested kinematics info is printed on the graph to make it easy to share it and avoid confusion.
 * Updated the way belt similarity and mechanical health are calculated on the belt comparison graph. Now it should be more "precise" -> easier to get 100% and avoid bullshit and chasing unicorns, but it can also be quite low if there are real mechanical problems on the machine. The mechanical health indicator should also be more accurate and give less false positives.
 * The ACCEL_PER_HZ is now a parameter for the macros that use it. So it can be changed at runtime and you don't have to change it in your config anymore. Also, it's now printed on the graphs so you can quickly see and remember which run was done at which accel per hz.
 * Paired peaks in belts comparison plots now use the Greek alphabet to avoid confusion with A and B belts, as was often the case in the past... The way the amplitude delta (in the small table) is calculated has also been updated to have the same reference in the calculation and to make it easier to compare one value to the other. This should make it more accurate and easier to interpret. 
 * Updated the way the input shaper filters are recommended (Performance Shaper vs. Low Vibration Shaper), as it was a bit too conservative since Shake&Tune v2.6, where I started including damping ratio and SCV in the calculation. Now it should be more correct: performance shaper should be preferred and low vibration should be used if the performance shaper is not enough to filter out ringing.
 * Added the PWM target freq when running TMC Autotune on the vibrations graph to tweak it and record its effects
 * Completely reworked the axes map calibration macro: now it finally works correctly and gives the correct orientation! It also generates a graph with a few more insights to diagnose if your accelerometer is working correctly.
 * Another new type of graph is the one that accompanies the static frequency excitation macro. This is optional, as your ears should be more than enough to find problematic components, but if you want to record things and compare the effect of this or that, you can do it. Be careful and read the corresponding documentation to avoid getting into the pitfalls of this recording, as it can be potentially misleading if not used correctly: that's why it's optional and your ears should be preferred first.

# Fixes

 * Because of all the new features, the documentation has been updated to reflect all the changes to the new graphs. There are also some additions like a suggested tuning workflow.
 * Better CSV file filtering to avoid including CSV files from other plugins in the /tmp folder. Now it should be more robust in this aspect.
 * Since Shake&Tune is now more tightly coupled with Klipper, it was easier to manage the accelerometers. So now the ACCEL_CHIP parameter is mostly not needed anymore and Shake&Tune should automatically select the right accelerometer for the test. You can still force one for the macros that still provide the parameter. This also has the side effect of fixing the use of non-standard accelerometer configurations with a custom name or multi-accelerometer configurations like bed swinger printers!

### CODE REVIEW

File: shaketune/graph_creators/axes_map_graph_creator.py [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+    new_coeffs = [pywt.threshold(c, threshold, mode='soft') for c in coeffs]
+    denoised_data = pywt.waverec(new_coeffs, wavelet)
+
+    # Compute noise by subtracting denoised data from original data
+    noise = data - denoised_data[: len(data)]
+    return denoised_data, noise
+
+
+def integrate_trapz(accel: np.ndarray, time: np.ndarray) -> np.ndarray:
+    return np.array([np.trapz(accel[:i], time[:i]) for i in range(2, len(time) + 1)])
```
[the remark this answers is not in the document]
Comment (Frix-x) [reply shaketune/graph_creators/axes_map_graph_creator.py:92]: Can you make a suggestion?

**Comment by celtare21** [comment]
```
$ AXES_SHAPER_CALIBRATION
// Shake&Tune version: v3.0.0-50-g6db1d39
!! Error on '_AXES_SHAPER_CALIBRATION FREQ_START=5 FREQ_END=133.33 HZ_PER_SEC=1 ACCEL_PER_HZ= AXIS=all SCV= MAX_SMOOTHING= TRAVEL_SPEED=120 Z_HEIGHT=': unable to parse

```
needed to manually pass all parameters for it to work:
```
$ AXES_SHAPER_CALIBRATION ACCEL_PER_HZ=75 SCV=5 Z_HEIGHT=20 MAX_SMOOTHING=0.1
// Shake&Tune version: v3.0.0-50-g6db1d39
// Testing frequency: 6 Hz
// Testing frequency: 7 Hz
```

**Comment by celtare21** [comment]
Same for COMPARE_BELTS_RESPONSES

**Comment by Frix-x** [comment]
> ```
> $ AXES_SHAPER_CALIBRATION
> // Shake&Tune version: v3.0.0-50-g6db1d39
> !! Error on '_AXES_SHAPER_CALIBRATION FREQ_START=5 FREQ_END=133.33 HZ_PER_SEC=1 ACCEL_PER_HZ= AXIS=all SCV= MAX_SMOOTHING= TRAVEL_SPEED=120 Z_HEIGHT=': unable to parse
> ```
> 
> needed to manually pass all parameters for it to work:
> 
> ```
> $ AXES_SHAPER_CALIBRATION ACCEL_PER_HZ=75 SCV=5 Z_HEIGHT=20 MAX_SMOOTHING=0.1
> // Shake&Tune version: v3.0.0-50-g6db1d39
> // Testing frequency: 6 Hz
> // Testing frequency: 7 Hz
> ```

Yes that's quite normal. I implemented a fix that works for Fluidd but not for Mainsail for now. I'm working on it. There's also a Jekyll documentation website that I plan to deploy before the release. There's still a couple things to fix like this :)

I just opened the PR to get the Sourcery review and prepare the final merge.

But I'll try to fix this issue today. I have some ideas to do it cleanly :)

**Comment by celtare21** [comment]
> > ```
> > $ AXES_SHAPER_CALIBRATION
> > // Shake&Tune version: v3.0.0-50-g6db1d39
> > !! Error on '_AXES_SHAPER_CALIBRATION FREQ_START=5 FREQ_END=133.33 HZ_PER_SEC=1 ACCEL_PER_HZ= AXIS=all SCV= MAX_SMOOTHING= TRAVEL_SPEED=120 Z_HEIGHT=': unable to parse
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > needed to manually pass all parameters for it to work:
> > ```
> > $ AXES_SHAPER_CALIBRATION ACCEL_PER_HZ=75 SCV=5 Z_HEIGHT=20 MAX_SMOOTHING=0.1
> > // Shake&Tune version: v3.0.0-50-g6db1d39
> > // Testing frequency: 6 Hz
> > // Testing frequency: 7 Hz
> > ```
> 
> Yes that's quite normal. I implemented a fix that works for Fluidd but not for Mainsail for now. I'm working on it. There's also a Jekyll documentation website that I plan to deploy before the release. There's still a couple things to fix like this :)
> 
> I just opened the PR to get the Sourcery review and prepare the final merge.
> 
> But I'll try to fix this issue today. I have some ideas to do it cleanly :)

Great, didn't know you were aware of this. Just a small comment, I am on Fluidd not Mainsail

**Comment by Frix-x** [comment]
> Great, didn't know you were aware of this. Just a small comment, I am on Fluidd not Mainsail

Actually I just tested it again and it should work correctly on both Mainsail and Fluidd... Are you sure you are on the latest commit from the develop branch?

**Comment by celtare21** [comment]
> > Great, didn't know you were aware of this. Just a small comment, I am on Fluidd not Mainsail
> 
> Actually I just tested it again and it should work correctly on both Mainsail and Fluidd... Are you sure you are on the latest commit from the develop branch?

Yep, just did a git pull (I'm on commit ecd57ea3dcf7791567cc04830b1755213dc6f54c), exactly the same error

**Comment by Frix-x** [comment]
> Yep, just did a git pull (I'm on commit [ecd57ea](https://github.com/Frix-x/klippain-shaketune/commit/ecd57ea3dcf7791567cc04830b1755213dc6f54c)), exactly the same error

Well... So I will need to do more testing and understand what is not working on your side. I'll probably try to solve this issue with a another method then...

**Comment by Frix-x** [comment]
@ozelentok I think beside the last small bugs mentionned above, I'm near a release for S&Tv4 based on your rework. If you want to do a review and/or have some comments about it, do not hesitate to share them here :)

**Comment by ozelentok** [comment]
Everything looks great :)
I also encountered the bug mentioned, turns out jinja uses a special Undefined object to mark missing arguments
```
dict_items([('FREQ_START', 5), ('FREQ_END', 133.33), ('HZ_PER_SEC', 1), ('ACCEL_PER_HZ', Undefined), ('AXIS', 'all'), ('SCV', Undefined), ('MAX_SMOOTHING', Undefined), ('TRAVEL_SPEED', 120), ('Z_HEIGHT', Undefined)]
```
Adding the `is defined` test to the filtering in `dummy_macros.cfg` solved it for me
```
if value is defined and value is not none and value != ''
```

**Comment by Frix-x** [comment]
> Everything looks great :) I also encountered the bug mentioned, turns out jinja uses a special Undefined object to mark missing arguments
> 
> ```
> dict_items([('FREQ_START', 5), ('FREQ_END', 133.33), ('HZ_PER_SEC', 1), ('ACCEL_PER_HZ', Undefined), ('AXIS', 'all'), ('SCV', Undefined), ('MAX_SMOOTHING', Undefined), ('TRAVEL_SPEED', 120), ('Z_HEIGHT', Undefined)]
> ```
> 
> Adding the `is defined` test to the filtering in `dummy_macros.cfg` solved it for me
> 
> ```
> if value is defined and value is not none and value != ''
> ```

Thanks, I pushed your suggestion!
I cannot replicate the bug on my side, not sure why... @celtare21 can you test and let me know if it's ok now?

**Comment by celtare21** [comment]
> > Everything looks great :) I also encountered the bug mentioned, turns out jinja uses a special Undefined object to mark missing arguments
> > ```
> > dict_items([('FREQ_START', 5), ('FREQ_END', 133.33), ('HZ_PER_SEC', 1), ('ACCEL_PER_HZ', Undefined), ('AXIS', 'all'), ('SCV', Undefined), ('MAX_SMOOTHING', Undefined), ('TRAVEL_SPEED', 120), ('Z_HEIGHT', Undefined)]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Adding the `is defined` test to the filtering in `dummy_macros.cfg` solved it for me
> > ```
> > if value is defined and value is not none and value != ''
> > ```
> 
> Thanks, I pushed your suggestion! I cannot replicate the bug on my side, not sure why... @celtare21 can you test and let me know if it's ok now?

Can confirm it works just fine now

**Pull request merged**